tests/Fixer/Phpdoc/PhpdocScalarBreakdownFixerTest.php

```diff
@@ -0,0 +1,293 @@
+ * @covers \PhpCsFixer\Fixer\Phpdoc\PhpdocScalarBreakdownFixer
+ */
+final class PhpdocScalarBreakdownFixerTest extends AbstractFixerTestCase
+{
+    public function testBasicFix()
+    {
+        $expected = <<<'EOF'
+<?php
+    /**
+     * @return int|string|float|bool
+     */
+
+EOF;
+
+        $input = <<<'EOF'
+<?php
+    /**
+     * @return scalar
+     */
+
+EOF;
+
+        $this->doTest($expected, $input);
+    }
+
+    public function testPropertyFix()
+    {
+        $expected = <<<'EOF'
+<?php
+/**
+ * @method int|string|float|bool foo()
+ * @property int|string|float|bool $foo
+ * @property-read int|string|float|bool $bar
+ * @property-write int|string|float|bool $baz
+ */
+
+EOF;
+
+        $input = <<<'EOF'
+<?php
+/**
+ * @method scalar foo()
+ * @property scalar $foo
+ * @property-read scalar $bar
+ * @property-write scalar $baz
+ */
+
+EOF;
+
+        $this->doTest($expected, $input);
+    }
+
+    public function testDoNotModifyVariables()
+    {
+        $expected = <<<'EOF'
+<?php
+    /**
+     * @param int|string|float|bool $scalar
+     */
+
+EOF;
+
+        $input = <<<'EOF'
+<?php
+    /**
+     * @param scalar $scalar
+     */
+
+EOF;
+
+        $this->doTest($expected, $input);
+    }
+
+    public function testFixWithTabsOnOneLine()
+    {
+        $expected = "<?php /**\t@return\tint|string|float|bool\t*/";
+
+        $input = "<?php /**\t@return\tscalar\t*/";
+
+        $this->doTest($expected, $input);
+    }
+
+    public function testFixMoreThings()
+    {
+        $expected = <<<'EOF'
+<?php
+    /**
+     * Hello there mr scalar!
+     *
+     * @param int|string|float|bool $scalar
+     *
+     * @return int|string|float|bool
+     */
+
+EOF;
+
+        $input = <<<'EOF'
+<?php
+    /**
+     * Hello there mr scalar!
+     *
+     * @param scalar $scalar
+     *
+     * @return scalar
+     */
+
+EOF;
+
+        $this->doTest($expected, $input);
+    }
+
+    public function testFixVar()
+    {
+        $expected = <<<'EOF'
+<?php
+    /**
+     * @var int|string|float|bool Some integer value.
+     */
+
+EOF;
+
+        $input = <<<'EOF'
+<?php
+    /**
+     * @var scalar Some integer value.
+     */
+
+EOF;
+
+        $this->doTest($expected, $input);
+    }
+
+    public function testFixType()
+    {
+        $expected = <<<'EOF'
+<?php
+    /**
+     * @type int|string|float|bool
+     */
+
+EOF;
+
+        $input = <<<'EOF'
+<?php
+    /**
+     * @type scalar
+     */
+
+EOF;
+
+        $this->doTest($expected, $input);
+    }
+
+    public function testDoNotFix()
+    {
+        $expected = <<<'EOF'
+<?php
+    /**
+     * @var scalara
+     */
+
+EOF;
+
+        $this->doTest($expected);
+    }
+
+    public function testComplexMix()
+    {
+        $expected = <<<'EOF'
+<?php
+    /**
+     * @var notabooleanthistime|int|string|float|bool|integerr
+     */
+
+EOF;
+
+        $input = <<<'EOF'
+<?php
+    /**
+     * @var notabooleanthistime|scalar|integerr
```

Comment:
please add test with `?scalar` and `Collection<scalar>`

Reply:
You mean `scalar[]`? AFAIK `Collection<>` is no valid PHPDoc type.

Reply:
No, I do mean `@var ?scalar` and `@var Collection<scalar>`, as stated.
both are valid according to psr-5 proposal, which we are following across the project.

Reply:
Oh, ok didn't even notice that one. Will do then

Reply:
AFAIK `@var ?scalar` is not even valid. If you add type hinting with `?string` you will get a `@param null|string`. So why should there be a test for it? Or should this be replaced with `null|int|string|float|bool` then?

Reply:
> I guess this should be done in the `PhpdocTypeFixer` to replace `?string` with `null|string`?

No, why ? How did you come up with this? I didn't asked for it, `?string` is perfectly valid, and since 7.1 it's better than `null|string` alternation, as it could by typehinted as-is in function prototype.

> `AFAIK @var ?scalar is not even valid`

It is. It is valid as much as `?string` is.

> If you add type hinting with `?string` you will get `@param null|string`

No, If I add typehinting `?foo` I will get `@param ?foo`. `?foo` is new way of saying type is nullable without saying what is default value, one could use it as-is in documentation.

> So why should there be a test for it? 

To prove it works properly.

> Or should this be replaced with `null|int|string|float|bool` then?

indeed

Reply:
> It is. It is valid as much as ?string is.
In which spec is this even listed?

Reply:
phpdoc accepts any type that is valid for php itself and it's combinations, like `int|string` or `int[]`.
`?string` is valid PHP type (it is a type, it is not a string with null as default value - similar, yet different)

Reply:
> Doesn't make sense to me.
> will be a shown as valid input. but it is not

typehint is not taken into consideration if it is only in typehint. for **any** type. 
not only for `?foo`. use it properly and you will see error
`Fatal error: Uncaught TypeError: Argument 1 passed to foo() must be of the type string or null, integer given`

https://3v4l.org/MbOgD

Reply:
> The new declaration for psr-5 `(int|bool|float|string)[]` is currently not valid for phpdoc.

why you say it is not valid? it is valid according to spec. if some IDE is not supporting it - issue of IDE.
yet, sure, if you want to do `int[]|bool[]|float[]|string[]`, go ahead

Reply:
Then if we are going to full phpspec. The `PhpdocTypeFixer` and `PhpdocScalarFixer` won't work with `?STRING` or anything else. because it is not listed as type inside theses classes.

So we have to rewrite the whole thing and consider `?` and `[]` as type modifier which should be ignored. Currently `PhpdocTypeFixer` simply cuts `[]` off and appends it again at the end after doing some fixes.

So this change will take some time to extract the modifiers of the type and pass em down to any specific type fixing itself.